### PR TITLE
Creation of basic roles to create custom rpi configuration

### PR DESCRIPTION
**Pull request by chloeskt**
Hi all, this huge PR aims at:

- creating a few basic roles which we will need to configure our fleet of RPI:
	- **password_deactivator**: disable password authentification 
	- **ssh_role**: does the creation and exchange of ssh keys on all hosts (including local machine) so as to permit ssh connexions and then reverse_ssh (another PR)
	- **rpi_camera_config**: enable camera usage on the RPI which are equipped of cameras (they are to be listed in `host_vars/hosts.yml` (which is the inventory) under the name `raspberry_with_camera`

- creating playbooks associated with each of these roles (might be a bit redundant):
    - **deactivate_password_playbook.yml**
    - **add_ssh_key_playbook.yml**
    - **enable_camera_playbook.yml**

- I also created some playbooks which are not attached to roles:
    - **common_playbook.yml:** common to all hosts, it is the raspberry configuration aiming at installing python, pip3, python3-venv, do of full upgrade of packages on the first boot only, install docker ARM (thanks to geerlingguy role) and create a folder to store images. This could become a role if wanted or when we have some stuff to install and create, different playbooks or roles.
    - **stop_service_playbook.yml**: to run when we want to quickly stop a service, for instance docker. 

Now more general comments on what else has been added:
- **requirements.yml:** needed when we want to import some roles from ansible-galaxy
- **requirements.txt:** requirements which will be installed on your local venv
- **rpi.requirements.txt:** requirements which will be installed on your fleet of RPI venv
- **ansible.cfg:** Ansible configuration file 
- **RpiSETUP.md**: to explain how to setup your RPI prior of the first boot, in order to make our all repo work (flash SD card, enable ssh and wifi connexion)
- **master.yml**: the master playbook  

This is a first version to show you what is possible with Ansible and introduce a few features which we might come in handy.  

Finally I’ve added to toy script, just so that you can have an example of how to run a python script using Ansible. It is in `scripts/take_picture.py`. This script is only run on RPI with camera (different from the main RPI which does not have a camera and its user and ip address need to be added in `.env`, see README.md for more information), it takes a picture every 5 seconds, 5 times and send each picture to the main RPI in the `/home/pi/images` folder. 

This script will be deleted afterwards, once we go deeper into the setup (adding pyro-engine, etc). It is just here now as a way for you to see how we can launch script and send files over rsync when the RPIs are on the same local network. 

As always, all reviews and feedback are welcome 🤗 Do you have any ideas of what also needs to be implemented to have the basic setup ready ? We also need to have a deeper look into the internet connexion setup and reverse ssh (in the separate PR I guess). I have already started to look at reverse ssh configuration (I am waiting on the choice of the CSP and the opening of a VPS to go further on this topic).

### CODE REVIEW

**Comment by fe51** [comment]
Hey there ! Thanks a lot @chloeskt for the PR (massive indeed), I will review it as sound as I can !

File: playbooks/common_playbook.yml
Comment (blenzi) [inline playbooks/common_playbook.yml]: small typo: commun -> common

File: host_vars/hosts.yml
Comment (blenzi) [inline host_vars/hosts.yml:18]: is there a way to avoid repeating the variables for rasp2?

File: roles/ssh_role/tasks/main.yml
Comment (blenzi) [inline roles/ssh_role/tasks/main.yml]: sorry, maybe I missed the logic: does it add the keys of all rpis as authorized keys to each of them?

**Review by blenzi — COMMENTED** [review]
Hi @chloeskt, thank you very much for this PR! I have never used ansible and only had a very quick look at the tutorial so my review cannot be thorough. Nevertheless the project looks well structured and documented, the steps look reasonable.

I only have two naive questions apart from the small comments:
- I probably missed why you need / use both the RSA and ECDSA keys ?
- how is this all tested?

File: playbooks/common_playbook.yml
Comment (chloeskt) [reply playbooks/common_playbook.yml]: done, thanks for spotting this !

File: roles/ssh_role/tasks/main.yml
Comment (chloeskt) [reply roles/ssh_role/tasks/main.yml]: This tasks adds the ecdsa keys of all rpis as known_hosts to each of them (this then allows each rpi to connect to all others. It is for ingoing connexion). The known_hosts file tells the rpi where it can go.

Whereas the task "set authorized_keys on rpi" adds the rsa keys of all rpis as authorized_keys to each of them (to allow all other rpis to connect to the other; this for outgoing connexion). The authorized_keys file tells the rpi which other rpi can come and ssh on it.

Comment (chloeskt) [reply roles/ssh_role/tasks/main.yml]: Maybe we can change the names of the tasks to make it more clear

File: host_vars/hosts.yml
Comment (chloeskt) [reply host_vars/hosts.yml:18]: Yes it seems that we can just specify the hostname `rasp2` without repearing the variables and it works.

**Comment by chloeskt** [comment]
Thank you @blenzi for your review and your questions. 

- **RSA** keys are used for **outgoing connexion.** 
- **ECDSA** keys are used for **ingoing connexion** (which is needed for instance in the example script, when the rpi with a camera takes a picture and send it to the main rpi using rsync, the ECDSA key is used. And this key is automatically created by the OS Raspbian). 

Regarding the tests, right now, I tested everything on my local machine and on my two rpi by running the master playbook. I do not know how we could setup a CI; maybe we could have a look at [molecule](https://molecule.readthedocs.io/en/latest/) but it seems complicated. Otherwise I wanted to setup a **lint check** in the CI, in another PR, using **yamllint** ! @fe51 also talked about it

If you want to test it on your local machine, you can modify the inventory: `host_vars/hosts.yml` with the right info of your rpis (if you have one with a camera and another one without). If you only have one with a camera for instance, you'll just have to comment out the rsync command in `take_picture.py` (otherwise the rpi will try to send a file to itself) and it should work. Or you can choose to only run part of the master playbook (for instance not the `playbooks/run_scripts_camera_playbook.yml` since its mainly here to provide an example) or only specific playbooks to check, as you wish !

File: RpiSETUP.md
Comment (fe51) [inline RpiSETUP.md]:
```suggestion
Then you can flash your SD Card with the chosen OS. see [this article from raspberrypi.org](https://www.raspberrypi.org/documentation/installation/installing-images/) to help you out writing the image.
```

Comment (fe51) [reply RpiSETUP.md]: let us make it smooth as possible

File: RpiSETUP.md
Comment (fe51) [inline RpiSETUP.md:1]: As I have just received a new pi, I have followed this set up in order to check it works !

File: RpiSETUP.md
Comment (fe51) [inline RpiSETUP.md]:
```suggestion
SSH needs to be enabled so that you can log onto your RPI. To do so, an empty ssh file must be created, on linux or mac os, you can run the following command:
```

It will be just a bit different on windows so let explain what we need

File: playbooks/common_playbook.yml
Comment (fe51) [inline playbooks/common_playbook.yml:35]: Do not we need a step/role to install geerlingguy.docker_arm ?  (a command like ansible-galaxy install eerlingguy.docker_arm ?)

File: roles/ssh_role/tasks/main.yml
Comment (fe51) [inline roles/ssh_role/tasks/main.yml:13]:
```suggestion

- name: create ssh directory
  file:
    path: ~/.ssh/
    state: directory

```
I did not have a ssh folder by defaut, thus it was failing ! (Might be a cleaner way to avoid this issue :) )

**Review by fe51 — COMMENTED** [review]
I have not investigated all aspects so far, but as far as will not be available this week end, I am adding some comments now ! 

On overall, I think this is fine and good start ! (maybe the more effcient way would be to quite quickly merge this PR and then provide correction/update quite quickly with small PRs :)

File: playbooks/common_playbook.yml
Comment (chloeskt) [reply playbooks/common_playbook.yml:35]: If you install the `requirements.yml` it should work

File: RpiSETUP.md
Comment (chloeskt) [reply RpiSETUP.md]: True, I forgot about windows, thanks !

File: roles/ssh_role/tasks/main.yml
Comment (blenzi) [reply roles/ssh_role/tasks/main.yml]: Sure, changing the name could clarify!

File: RpiSETUP.md
Comment (chloeskt) [reply RpiSETUP.md]: You're right, done

**Comment by blenzi** [comment]
Thanks for the replies @chloeskt ! It is still not clear to me why incoming and outgoing connections must use different keys but it is not an issue either.

**Comment by chloeskt** [comment]
> Thanks for the replies @chloeskt ! It is still not clear to me why incoming and outgoing connections must use different keys but it is not an issue either.

@blenzi 

I guess, to simplify we could do the following: only use the ECDSA key (the most secure key, which is why it is used for incoming connections. I believed it is linked to strict host checking - by default it is looking for the most secure ssh key on the server which is ECDSA on rpi because it is created by default. We could use a RSA key but we would need to remove the ECDSA key before) created by default on each rpi (in the folder `/etc/ssh/`).

In the `ssh_role` role we generate a rsa key (then they are all put in each other `authorized_keys` file to allow outgoing connection) but instead we could fetch the ecdsa key in `/etc/ssh/` and use this key in the `authorized_keys`. ECDSA keys can also be used for outgoing connection, I used the RSA one here because I created one at the beginning of the role.

When I started to think about this exchange of keys, by default I only fought about using RSA key (I guess it's the most common). I discovered ECDSA key only very recently (when debugging the rsync issue), thus I did not think about changing the code, but it might be easier just to use the ECDSA key created by default on each rpi for outgoing connections also. What do you think ? (don't know if it's clear enough, it's a new subject for me 😅)

File: roles/ssh_role/tasks/main.yml
Comment (chloeskt) [reply roles/ssh_role/tasks/main.yml:13]: Oh my bad ! I did not think about this case, nice catch, thanks 👌

File: playbooks/common_playbook.yml
Comment (chloeskt) [reply playbooks/common_playbook.yml:35]: the command is `ansible-galaxy install requirements.yml` I should've added this step in the README.md !

File: roles/ssh_role/tasks/main.yml
Comment (chloeskt) [reply roles/ssh_role/tasks/main.yml:13]: Thought the ssh folder is not in the RPI only if the RpiSETUP.md guidelines were not followed (because you need to created the .ssh)

But is it still cleaner to create the folder, I added this task

**Comment by blenzi** [comment]
Thanks @chloeskt ! I was also not aware of ECDSA keys :) I tend to think that if they are the most secure and are already available, why not using them all over. Again, no need to block the PR for that, I let you decide.

**Review by fe51 — APPROVED** [review]
Following our various exchanges, although there will soon be updates (ECDSA, etc.), I think that the most effective way to proceed is to merge this PR.

(Don't forget to delete the branch after merging :) )

**Pull request merged**
